xbmc/windowing/gbm/GBMUtils.cpp

```diff
@@ -0,0 +1,555 @@
+  {
+    CLog::Log(LOGERROR, "CGBMUtils::%s - failed to create surface", __FUNCTION__);
+    return false;
+  }
+
+  CLog::Log(LOGDEBUG, "CGBMUtils::%s - created surface with size %dx%d", __FUNCTION__,
+                                                                         m_drm->gbm->width,
+                                                                         m_drm->gbm->height);
+
+  return true;
+}
+
+void CGBMUtils::DestroyGbm()
+{
+  if(m_drm->gbm->surface)
+  {
+    gbm_surface_destroy(m_drm->gbm->surface);
+  }
+
+  m_drm->gbm->surface = nullptr;
+}
+
+bool CGBMUtils::SetVideoMode(RESOLUTION_INFO res)
+{
+  GetMode(res);
+
+  gbm_surface_release_buffer(m_drm->gbm->surface, m_bo);
+
+  m_bo = gbm_surface_lock_front_buffer(m_drm->gbm->surface);
+  m_drm_fb = DrmFbGetFromBo(m_bo);
+
+  auto ret = drmModeSetCrtc(m_drm->fd,
+                            m_drm->crtc_id,
+                            m_drm_fb->fb_id,
+                            0,
+                            0,
+                            &m_drm->connector_id,
+                            1,
+                            m_drm->mode);
+
+  if(ret == -1)
+  {
+    CLog::Log(LOGERROR,
+              "CGBMUtils::%s - failed to set crtc mode: %dx%d%s @ %d Hz",
+              __FUNCTION__,
+              m_drm->mode->hdisplay,
+              m_drm->mode->vdisplay,
+              m_drm->mode->flags & DRM_MODE_FLAG_INTERLACE ? "i" : "",
+              m_drm->mode->vrefresh);
+
+    return false;
+  }
+
+  CLog::Log(LOGDEBUG, "CGBMUtils::%s - set crtc mode: %dx%d%s @ %d Hz",
+            __FUNCTION__,
+            m_drm->mode->hdisplay,
+            m_drm->mode->vdisplay,
+            m_drm->mode->flags & DRM_MODE_FLAG_INTERLACE ? "i" : "",
+            m_drm->mode->vrefresh);
+
+  return true;
+}
+
+bool CGBMUtils::GetMode(RESOLUTION_INFO res)
+{
+  m_drm->mode = &m_drm_connector->modes[atoi(res.strId.c_str())];
+
+  CLog::Log(LOGDEBUG, "CGBMUtils::%s - found crtc mode: %dx%d%s @ %d Hz",
+            __FUNCTION__,
+            m_drm->mode->hdisplay,
+            m_drm->mode->vdisplay,
+            m_drm->mode->flags & DRM_MODE_FLAG_INTERLACE ? "i" : "",
+            m_drm->mode->vrefresh);
+
+  return true;
+}
+
+void CGBMUtils::DrmFbDestroyCallback(struct gbm_bo *bo, void *data)
+{
+  struct drm_fb *fb = static_cast<drm_fb *>(data);
+
+  if(fb->fb_id)
+  {
+    drmModeRmFB(m_drm->fd, fb->fb_id);
+  }
+
+  delete (fb);
+}
+
+drm_fb * CGBMUtils::DrmFbGetFromBo(struct gbm_bo *bo)
+{
+  {
+    struct drm_fb *fb = static_cast<drm_fb *>(gbm_bo_get_user_data(bo));
+    if(fb)
+    {
+      return fb;
+    }
+  }
+
+  struct drm_fb *fb = new drm_fb;
+  fb->bo = bo;
+
+  uint32_t width = gbm_bo_get_width(bo);
+  uint32_t height = gbm_bo_get_height(bo);
+  uint32_t stride = gbm_bo_get_stride(bo);
+  uint32_t handle = gbm_bo_get_handle(bo).u32;
+
+  auto ret = drmModeAddFB(m_drm->fd,
+                          width,
+                          height,
+                          24,
```

Comment:
If you use drmModeAddFB2, then you can pass the format directly and avoid the hardcoded 24/32. This would also allow you to support, e.g., 10bpc formats.